### ISSUE TITLE
MLP-Mixer: An all-MLP Architecture for Vision

### ISSUE THREAD

**Issue by Chicoryn** [Chicoryn/dream-go#55]
https://arxiv.org/pdf/2105.01601.pdf

> Convolutional Neural Networks (CNNs) are the go-to model for computer vision. Recently, attention-based networks, such as the Vision Transformer, have also become popular. In this paper we show that while convolutions and attention are both sufficient for good performance, neither of them are necessary. We present MLP-Mixer, an architecture based exclusively on multi-layer perceptrons (MLPs). MLP-Mixer contains two types of layers: one with MLPs applied independently to image patches (i.e. "mixing" the per-location features), and one with MLPs applied across patches (i.e. "mixing" spatial information). When trained on large datasets, or with modern regularization schemes, MLP-Mixer attains competitive scores on image classification benchmarks, with pre-training and inference cost comparable to state-of-the-art models. We hope that these results spark further research beyond the realms of well established CNNs and Transformers.

**Comment by Chicoryn** [comment]
![image](https://user-images.githubusercontent.com/77159/120896582-8ce5f300-c622-11eb-9c2d-582823774659.png)

Initial results with pure MLP-Mixer (without SAM) and the following parameters is not very promising. We also had to decrease the batch size to _128_ in order to fit in the memory of our GPUs:
- `embeddings_size` is _722_
- `tokens_mlp_dims` is _256_
- `channels_mlp_dims` is _2048_
- `num_blocks` is _9_
- `tile_size` is _5_ (with overlapping tiles, for a total of 49 tiles)

## Analysis

Based on the initial attempts we see several challenges we need to address in the architecture:
- The `embeddings_size` needs to be very big due to the global average pooling used to get the intermediate representation. This causes runtime performance and memory problems.
- The _value_ head does not seem to converge despite the _ownership_ head converging relatively well.
- It is not clear how `tile_size` should be tuned, we need to experiment.

### Number of parameters

Another problem is that MLP-Mixer models are huge in comparison to the ResNet models currently deployed (and does not produce comparable results). The example graphs in the beginning of this post has an MLP-Mixer model that is approximately 11 times bigger than the ResNet model.

| # | ResNet | MLP Mixer |
| --- | --- | --- |
| 9 | 9 · (2 · `filter_size` · `filter_size` · `num_channels` · `num_channels`) = 2,654,208 | 9 · (2 · `embeddings_size` · `tokens_mlp_dims` + 2 · `embeddings_size` · `channels_mlp_dims`) = 29,942,784 |

**Comment by Chicoryn** [comment]
See Figure 1 for the final measurements for different tile sizes, when using a stride of `tile_size // 2 + 1`. Results are the same as before, pretty damn bad. Oddly enough the _ownership_ head has a much higher accuracy than the _value_ head, but it does not transfer for some reason:

| `tile_size` | `3` | `5` | `7` | `9` | `13` | `15` |
| --- | --- | --- | --- | --- | --- | --- |
| Policy (Top 1) | 16.5% | 18.5% | **27.5%** | 27.1% | 16.6% | 21.3% |
| Policy (Top 3) | 30.1% | 33.7% | 46.0% | **46.7%** | 30.5% | 38.2% |
| Policy (Top 5) | 40.0% | 41.8% | 55.3% | **55.7%** | 39.4% | 46.9% |
| Value | 53.0% | 55.0% | 53.7% | 54.2% | **54.3%** | 52.6% |

**Comment by Chicoryn** [comment]
An alternative approach, inspired by Axial-SWideRNet [1], would be to just add a single block at the end of the convolutional residual network. This would serve as a _global attention_ block just before the _value_ and _policy_ head.

The main drawback of MLP-Mixer blocks is that they require a lot of parameters, when computing such a _global attention_ block to keep the same dimensions as a similar residual block:
* `embeddings_size` is `128` (generally referred to as `num_channels` in our existing code base)
* `tokens_mlp_dims` is `361`
* `channels_mlp_dims` is `128`

This yields a parameter count of `125,184` (compared to `294,912` for a single residual block). This should be feasible to add at the end of the convolutional stem. Alternatively after every `n` residual blocks.

[1] Axial-DeepLab: StandAlone Axial-Attention for Panoptic Segmentation, https://arxiv.org/abs/2003.07853

**Comment by Chicoryn** [comment]
I had to tweak the `max_learning_rate` to `0.001` (from `0.01`) because the MLP mixer collapsed with the higher learning rate. There is probably some hyper-parameter tuning that could be done here to improve the results, so take that in mind when reviewing the test scores:

| | ref | at end of stem | every `1` residual block |
| --- | --- | --- | --- |
| Policy (Top 1) | 53.1% | 51.8% | x% |
| Policy (Top 3) | 76.9% | 75.9% | x% |
| Policy (Top 5) | 85.3% | 84.3% | x% |
| Value              | 67.4% | 66.8% | x% |

**Comment by Chicoryn** [comment]
Overall adding MLP Mixer blocks to the stem shows good results, but there is a strong anomaly with the _value_ head. All metrics except for it is converging properly, but it is not. Based on the training accuracy this is probably due to overfitting. Pink is the training accuracy, and green is the evaluation accuracy:

![image](https://user-images.githubusercontent.com/77159/124656181-27935500-dea1-11eb-8cb8-0f6a1b26f074.png)

**Comment by Chicoryn** [comment]
From a brief performance benchmark of a cuDNN implementation of a single MLP Mixer block, in addition to bottleneck blocks [1], compared to an standard residual block we see that they are pretty similar runtime performance. We might be able to squeeze out some additional performance for the mixer blocks by using a dedicated BLAS library for the dense layers instead of mimicking it using a 1x1 convolution:

```
test layers::bottleneck_block::tests::bottleneck_block ... bench:     338,165 ns/iter (+/- 15,251)
test layers::mixer_block::tests::mixer_block           ... bench:     425,699 ns/iter (+/- 20,147)
test layers::residual_block::tests::residual_block     ... bench:     480,611 ns/iter (+/- 114,238)
```

These numbers would give the following comparative performance when put together into a full stem. So far the bottleneck architecture seems to be performance worse than the others (private run), but a `5x` stem of residual and mixer blocks might be attractive.

| Configuration | Runtime (ns) |
| --- | ---: |
| `9x` residual | `4,325,499` |
| `5x` residual + mixer | `4,531,550` |
| `6x` residual + mixer | `5,437,860` |
| `6x` bottleneck+ mixer | `4,583,184` |

**Comment by Chicoryn** [comment]
I don't see any meaningful improvement combining residual and MLP blocks during training. I've tried a few different combination to see if we get any different results where `R` represents a residual block and `M` an MLP-Mixer block:

| Configuration | Policy (%) | Value (%) |
| --- | ---: | ---: |
| `RMRMRMRMR` | 52.8% | 65.5% |
| `RRMRRMRRM` | 52.8% | 66.1% |
| `MMMMMMRRR` | 51.3% | 63.8% |
| `RRRRRRMMM` | xx.x% | xx.x% |
